### ISSUE TITLE
Docker support would be great

### ISSUE THREAD

**Issue by stevenw** [openai/spinningup#3]

**Comment by machinaut** [comment]
Could you describe the use case or what features you'd like that are missing?

**Comment by xeb** [comment]
One challenge is going to be the MuJoCo license & installation. Some of the required files (e.g. in ~/.mujoco) can be mapped with a volume. If you will accept a PR for a Dockerfile, that would be appreciated. If not, we can create a spinningup-docker repo as an example. I'm interested in helping make this work.

**Comment by machinaut** [comment]
I think this would be a good solution if you could run all of the code in it.

@xeb I'm curious how you would get the plotting or rendering functionality to work.  In the past it has been difficult to get docker containers to render graphical interactive windows.

**Comment by machinaut** [comment]
@xeb keep us updated!  If it ends up working then it would be a good thing to have in the repo!

**Comment by gvgramazio** [comment]
I'm interested in contributing in this feature. What are the requirements of the feature?

**Comment by machinaut** [comment]
I think I would be most interested in merging something that allowed full functionality via docker (instead of only partial).

The tricky part here is getting the plotting to function as expected.

**Comment by gvgramazio** [comment]
The only thing I can't testing is MuJoCo license because I don't have any. If it's fully functional except for the MuJoCo license (that could be implemented by anyone that have it) then are you interested?
Could you explain better the plotting functionality? Are you interested in plotting or in visualizing the plot directly in the container? The Docker container should have support for any other library/tool not explicitly listed as required?

**Comment by cirocavani** [comment]
I create a Dockerfile that "pass" the installation test (includes the plot tool, but no MuJoCo). It also uses NoVNC javascript VNC client to render X on a browser directly.

https://github.com/cirocavani/reinforcement-learning/tree/master/openai-spinningup

**Comment by gvgramazio** [comment]
@cirocavani I've taken a look at your repo. There are a few suggestions:

- You should never use things like `upgrade`, `dist-upgrade` and similar. One of the major points of using Docker, that is have a control on the version of packages used, would be wasted.
- There is no point in using conda in a Docker container. You are not going to mix python versions.
- There are better ways of importing git repositories. The way you are using is wrong. Either specify the commit or add a random variable to your image before the download. Right now, if there is any change in the git repo, your image will be not up to date.

**Comment by gvgramazio** [comment]
@cirocavani I forgot another suggestion: do not use sudo inside Docker images.

**Comment by gvgramazio** [comment]
I've written a basic dockerfile with the requirements stated so far (is capable of running examples and plotting results) to be included in the main project. See [my fork](https://github.com/gvgramazio/spinningup/tree/docker) to try it.

In order to work with it, simply build the image:

    sudo docker build -t spinningup .

Adjust the permissions of xhost:

    xhost +local:root

Run the container exposing the xhost so that container can render to the correct display by reading and writing through the X11 Unix socket:

    sudo docker run -it --rm \
      -e DISPLAY=$DISPLAY \
      -v /tmp/.X11-unix:/tmp/.X11-unix \
      spinningup

Enjoy your container. Of course, you can plot results.

When have you done with it, adjust again the permissions of xhost:

    xhost -local:root

---

Note that I'm writing this post in order to receive a feedback and feature requests. I don't consider it ready for a pull request yet.

I will be satisfied by adding the support for python (currently works only with python3) and managing in a better and safer way the x11 redirection. Both things should be done by tomorrow. Feel free to ask for more features.

**Comment by machinaut** [comment]
December is almost over (at least from a work-week perspective)!  @cirocavani thanks for the code and I'll have time to look at it in the new year.
Really appreciate your hard work on this one!

**Comment by alek5k** [comment]
Hello all,

I decided I wanted to work on a virtual machine rather than using docker (to avoid issues with x11 and xhost, and to leverage GPU if needed), so I wrote a bash setup script that installs SpinningUp and all dependencies from scratch. You can find it here:

https://gist.github.com/alek5k/dbd6c7235c46584bd25221c52f27d308

The script is mostly unattended, you just need to enter sudo user and pass once to install apt packages.
It makes some assumptions for where you want to set up your folders, but i've flagged that as a 'todo' item to make it configurable in the script if it becomes popular.

I added this here in-case this information can be used in future for the 'automation' aspect of setting up SpinningUp (not limited to docker, perhaps for use with vagrant, AWS/Azure VMs, etc)

I tested it on a clean Ubuntu 18.04.1 x64 virtual machine.

**Comment by gvgramazio** [comment]
You can both use x11 and GPU with docker (as long as you have an Nvidia). Working on a docker container instead of a virtual machine could be better for performance, portability and testing. These are the reasons why a docker support would be great. Unfortunately I hadn't a feedback from the developers of spinningup in the last month.

**Comment by thomasarvanitidis** [comment]
@gvgramazio, is your approach only working with docker running on a linux host? I'm can't plot or display the results of [PPO in the LunarLander-v2](https://spinningup.openai.com/en/latest/user/installation.html#check-your-install).
Specifically, `python3 -m spinup.run plot /spinningup/data/installtest/installtest_s0` breaks with:
```
_tkinter.TclError: couldn't connect to display "/private/tmp/com.apple.launchd.uuzCe1hm7B/org.macosforge.xquartz:0"
```

**Comment by gvgramazio** [comment]
@thomasarvanitidis Yes, it's supposed to work only on a linux host. Sorry about that.

**Comment by erensunerr** [comment]
I think an official docker image would be amazing. I have trouble running spinning up on m1 mac :(
